ui/src/message_popup/popup_question.jsx

```diff
@@ -54,25 +59,53 @@ export default React.createClass({
   },
 
   onTextChanged({target:{value}}:TextChangeEvent) {
-    this.setState({ responseText: value })
+    this.setState({ initialResponseText: value })
   },
 
-  onSendPressed() {
-    const {elapsedMs, responseText} = this.state;
-    const {question, shouldShowStudentCard, allowedToToggleHint} = this.props;
+  onSavePressed() {
+    const {question, shouldShowStudentCard, allowedToToggleHint, shouldShowFeedback} = this.props;
+    this.logData();
+    if(shouldShowFeedback){
+      this.setState({isRevising:true});
+    }else{
+      this.props.onDone();
+    }
+  },
+  
+  onDonePressed(text) {
+    this.logData(text);
+    this.props.onDone();
+  },
+  
+  logData(finalText) {
+    const {elapsedMs, initialResponseText} = this.state;
+    const {question, shouldShowStudentCard, allowedToToggleHint, shouldShowFeedback} = this.props;
     const response:Response = {
       question,
       shouldShowStudentCard,
       allowedToToggleHint,
+      shouldShowFeedback,
       elapsedMs,
-      responseText
+      initialResponseText,
+      finalText
+    };
+    if(!this.state.isRevising){
```

Comment:
I'd suggest moving this to separate methods, maybe `logResponse()` and `logRevision(finalText)` and `logRevisionDeclined`.  Those methods can then call the more general `logData` method if there are bits you want to re-use.  That way it's explicit when you call the method as to what will be logged, rather than having that be determined within `logData`.